[Clang][Driver] Expose relocation model as multilib flags

statham-arm opened the pull request:
If a multilib collection contains libraries built for different methods of accessing global data (via absolute address, or via a GOT in -fPIC style, or as an offset from a fixed register in Arm -frwpi style), then `multilib.yaml` will need to know which relocation model an application is using in order to select the right library.

Even if a multilib collection only supports one relocation model, it's still useful for `multilib.yaml` to be able to tell if the user has selected the right one, so as to give a useful error message if they haven't, instead of silently selecting a library that won't work.

In this commit we determine the PIC / ROPI / RWPI status using the existing logic in `ParsePICArgs`, and translate it back into a canonical set of multilib selection flags.

statham-arm committed:
[Clang][Driver] Expose relocation model as multilib flags

If a multilib collection contains libraries built for different
methods of accessing global data (via absolute address, or via a GOT
in -fPIC style, or as an offset from a fixed register in Arm -frwpi
style), then `multilib.yaml` will need to know which relocation model
an application is using in order to select the right library.

Even if a multilib collection only supports one relocation model, it's
still useful for `multilib.yaml` to be able to tell if the user has
selected the right one, so as to give a useful error message if they
haven't, instead of silently selecting a library that won't work.

In this commit we determine the PIC / ROPI / RWPI status using the
existing logic in `ParsePICArgs`, and translate it back into a
canonical set of multilib selection flags.

statham-arm commented:
Note to reviewers: the thing I'm most uncertain of in this patch is whether I'm using the right strategy for finding out the information about relocation model.

`ParsePICArgs` looked sensible, but then failed an assertion if there was no "effective triple" set, so I dug out what _looked_ like the obvious way to figure one out, and did that.

But I'm not _very_ confident of what an "effective triple" is at all, and I don't know for sure that it's right to call `ComputeEffectiveClangTriple(Args)` to make one. So if there's any reason why that will give the wrong answer in a particular situation (e.g. because some information it depends on isn't finalized yet) then please don't assume I have a clever reason for knowing why it's safe after all :slightly_smiling_face:

statham-arm committed:
Add missing end of sentence in comment

(Thanks to clang-format for pointing out the trailing space on that
line, which was there because I meant to type a word after it!)

vhscampos commented on clang/lib/Driver/ToolChain.cpp:
```diff
@@ -376,6 +378,41 @@ ToolChain::getMultilibFlags(const llvm::opt::ArgList &Args) const {
+    // expect never to find out they're enabled. But it seems confusing to add
+    // -fno-ropi and -fno-rwpi unconditionally to every other architecture's
+    // multilib flags, so instead we leave them out completely.
+    if (IsARM) {
```
Is there any way to move this Arm-only segment into `getARMMultilibFlags`? In theory, that function handles everything that is only relevant to Arm. So it's a more natural place to host this segment here.

statham-arm commented:
> But I'm not _very_ confident of what an "effective triple" is at all, and I don't know for sure that it's right to call `ComputeEffectiveClangTriple(Args)` to make one.

Answering my own question: I had failed to spot that `ToolChain::getMultilibFlags` was _already_ calling `ComputeEffectiveClangTriple` at the top of the function! So I'm now more confident that it's sensible, but also, I could have just reused the existing value :slightly_smiling_face:

statham-arm committed:
Move ROPI/RWPI into the Arm-specific function

statham-arm committed:
Fix the test

All the CHECK lines are checked in a single run of FileCheck, so the
options they mention must appear in the same order that the flags are
shown in the -print-multi-flags output, which is ASCII sorting order.
